### ISSUE TITLE
slow download speed

### ISSUE THREAD

**Issue by mfolnovic** [madiele/vod2pod-rss#72]
app version: 1.0.0

- [ ] twitch api key set
- [x] youtube api key set

______
issue description: When i either download through browser, through podcast addict or if I copy/paste ffmpeg command and run it locally, the download speed is ~35KB/s. My internet speed is 1000/500 (Mbit/s).

Is there a configuration I'm missing?

Thanks for really useful product!

```
2023-05-12T04:52:36.239Z INFO  [app] app version 1.0.0
2023-05-12T04:52:36.274Z INFO  [app] starting server at http://0.0.0.0:8080
2023-05-12T04:52:36.274Z INFO  [actix_server::builder] starting 4 workers
2023-05-12T04:52:36.274Z INFO  [actix_server::server] Actix runtime found; starting in Actix runtime
2023-05-12T04:52:43.117Z INFO  [app] processing transcode at 192k for https://www.youtube.com/watch?v=BMr7Bz8eGus
2023-05-12T04:52:43.117Z INFO  [vod2pod_rss::transcoder] detected youtube url, need to find the stream url if not in cache
2023-05-12T04:52:46.076Z INFO  [vod2pod_rss::transcoder] generated ffmpeg command:
ffmpeg -ss 0 -i https://rr4---sn-8p2oxu-15bs.googlevideo.com/videoplayback/... -acodec libmp3lame -ab 192k -f mp3 -bufsize 5760 -maxrate 5760k -timeout 300 -hide_banner -loglevel error pipe:stdout
2023-05-12T04:52:46.080Z INFO  [vod2pod_rss::transcoder] streaming to client
```

**Comment by madiele** [comment]
I tried measuring the output of the raw ffmpeg command and yes, it seems to never go behind 40KB/s, I'll have to check if there is some kind of option I'm missing for ffmpeg or that is just how fast you can go, I'll let you know if I found a fix

**Comment by madiele** [comment]
seems to be youtube that is throttling, if you check in the yt-dlp github you'll see some mentions of it, we'll have to wait for a new yt-dlp version to come out with the fix. In the mean time it will be slow...

I suggest you watch the repo for new releases to get notified when an update comes out, I will also update all the dependencies, can't say if that will work but there is a chance.

Sorry I could not help.

**Comment by madiele** [comment]
I just added automatic pull requests with depend-a-bot when yt-dlp gets updated so that when they release a new version I get notified and i can make a new image quickly that hopefully fixes the issue✌️

**Comment by madiele** [comment]
The beta image now has a new yt-dlp version, check if that has improved something

Just add :beta to the end of the image section of the docker-compose file

**Comment by mfolnovic** [comment]
While I'm running :latest, it works (although slow).

When I update to :beta and try to download a podcast, I get an error in response "Exec format error (os error 8)".

Logs:

```
2023-06-25T14:44:34.280Z INFO  [app] app version 1.0.3-beta-5a25fe9
2023-06-25T14:44:34.298Z INFO  [app] detected version change (1.0.2 != 1.0.3-beta-5a25fe9) flushing redis DB
2023-06-25T14:44:34.299Z INFO  [app] starting server at http://0.0.0.0:8080
2023-06-25T14:44:34.299Z INFO  [actix_server::builder] starting 4 workers
2023-06-25T14:44:34.299Z INFO  [actix_server::server] Actix runtime found; starting in Actix runtime
2023-06-25T14:44:39.717Z INFO  [app] processing transcode at 192k for https://www.youtube.com/watch?v=X2RdwmPqBvQ
2023-06-25T14:44:39.718Z INFO  [vod2pod_rss::transcoder] detected youtube url, need to find the stream url if not in cache
```

If it helps, I'm running this on unraid.

**Comment by madiele** [comment]
seems like a new bug, I'll see if I can fix it

**Comment by madiele** [comment]
@mfolnovic should be fixed now

**Comment by jclendineng** [comment]
Can we re-open this? There are options we can use in yt-dlp to tweak the speeds, can we add yt-dlp options through here or would we need to compile ourselves with the options?

https://github.com/yt-dlp/yt-dlp/issues/10443 has some useful options that seem to fix the slowness issues.

--throttled-rate 500K -R infinite --socket-timeout 1
or similar options

**Comment by madiele** [comment]
Interesting, I'll reopen sure ✌️

As for when I'll find the time to look into this, no clue, but an option to add extra args to yt-dlp is not a bad idea and should be easy to add, if anyone wants to take a shot at it look at contributing.md for a guide on how to setup the dev environment, in open to PRs

**Comment by primeapple** [comment]
While I couldn't change the DL Speed, I had some success with changing the size of the downloaded audio.

Setting the following environment variable:

```
      - YOUTUBE_YT_DLP_GET_URL_EXTRA_ARGS=["-f", "bestaudio[format_note~=original]/bestaudio", "-S", "+size"]
```

This allows to sort by the smallest filesize available by YT. This is only the `-S` part. The other is to prefer the original language in case there are multiple language audiotracks available.

Maybe it helps someone.

**Comment by jclendineng** [comment]
Can we get aria2c bundled with this so we can use parallel downloads? So we can use it in the extra args value.

YOUTUBE_YT_DLP_GET_URL_EXTRA_ARGS:
["-x", "-N", "--downloader aria2c --downloader-args 'aria2c:--continue --max-concurrent-downloads=30 --max-connection-per-server=16 --split=30 --min-split-size=1M'"]

**Comment by madiele** [comment]
I don't thing that would work since vod2pod does not use yt-dlp to download, since it needs to support seek in podcast apps I use yt-dlp to get the stram url, then I use ffmpeg directly to download and convert it to mp3 on the fly, yt-dlp does not download anything from the

**Comment by kanjieater** [comment]
I'm also someone who's coming here because they use audiobookshelf, which then takes multiple days to download a few episodes kb at a time. Which, after days of waiting, sometimes it will get hung or stuck (haven't figured out if that's a vod2pod or abs issue yet). 

Could we have an option to just download the whole video saving it to tmp storage and serve it? yt-dlp can probably have it served in seconds if it downloads instead of days.

**Comment by Marcool04** [comment]
My understanding is that, basically, this is a limitation of vod2pod-rss design: it streams the audio, on demand, from youtube, via ffmpeg, to the client. There is no distinction between streaming, and downloading, so when a client tries to download a file, it still just gets a "trickle" from ffmpeg's stream and it takes a long time, or maybe times out depending on the file size.

Various workarounds have been suggested, but ultimately, fixing this "properly" requires vod2pod detecting, somehow, what the client wants: a stream, or a download? I'm not sure there is really any way to do that with an rss podcast, is there?

Or else it would require moving away from the ffmpeg streaming model, and using maybe, a temporary local file (which could be in RAM, as in /tmp?) which is created on the fly when download or stream is started, and can then be passed to the client as fast as desired, either in a trickle to be streamed, or as a download. So this would go:

- client request stream X (we don't know if it's a download or a stream)
- vod2pod uses yt-dlp to download audio file to /tmp and waits for it to fully download
- vod2pod kicks off ffmpeg and feeds it that file, with the output going to a URL
- vod2pod passes that ffmpeg output URL to client
- client receives the data, and can get it as fast as it wants.

Does that make sense?

**Comment by samuelvisser** [comment]
> Does that make sense?

Yeah it does. Ur right this is inherent to the design.

In my case its actually Vod2Pod which convinced me the download model is much better than trying to stream audio off of a video platform. I use Pinchflat for that now and auto- delete downloads after some time.

I think ur right that Vod2Pod would basically need to do the same, but im guessing that goes kinda fundamentally against the idea of this project, if i understand the creator right. There is download tools plenty, Vod2Pod is for when streaming works 'good enough'

**Comment by Marcool04** [comment]
I see what you mean @samuelvisser that this might be out of scope for vod2pod, but then again if it "converts a YouTube or Twitch channel into a podcast with ease" as its tagline says, then I don't know of any podcast clients that don't have an option to download an audio file to listen to later. So if vod2pod doesn't support that basic feature of podcasts, I feel that means that it only partially fulfills its own aim...

**Comment by madiele** [comment]
Yeah correct, I made vod2pod cause I mostly used to stream episode anyway (I live in Italy where data plans are extremely cheap). 
@Marcool04 is correct by design vod2pod is as lean as possible on storage footprint, I wanted it to work on a raspberry pi with little to no storage... Fundamentally implementing pre-fetching episodes will always look like an hack in vod2pod, so I don't know if I'll ever get around to do it. But yeah the best way to implement something like this would be to cache episodes while steaming and use an external cron job to trigger the download, but that it's not really an easy thing to implement and is full of egecases for sure. Any other way would probably require vod2pod to stray away from its design too much.

I'll be honest though, at the moment I don't have the free time to even try to implement something like this, sorry

**Comment by Marcool04** [comment]
I hear you @madiele , thanks for taking the time to respond, that all makes sense.
I don't think a change to first download the file to /tmp then stream it as vod2pod currently does would be too hard to implement, and I might take a stab at it if I get a bit of time some day soon. Would you be open to a PR along those lines?

(I have looked at the code and it makes sense to me, but maybe I am being too optimistic and I'll find stuff that makes it more complicated than I though once I start...)

**Comment by madiele** [comment]
Go for it, just maybe gave me a general idea on how you had in mind it would work, like if you need new endpoints, how do you trigger the preftech, how do you clean the tmp folder an so on, so I if you do end up making a PR it will be smooth sailing.

Of course I'm available to answer any questions, be sure to check the contributing.md of the repo, there are some tips on how to develop vod2pod
